### PR TITLE
Pr muzero categorical loss for reward value

### PR DESCRIPTION
**Pull request by emailweixu**

### CODE REVIEW

File: alf/utils/losses.py [hunk 1]
```diff
@@ -102,3 +104,371 @@ def multi_quantile_huber_loss(quantiles: torch.Tensor,
...
+            loss with the same shape as target
+        """
+        raise NotImplementedError()
+
+    def calc_expectation(self, pred: torch.Tensor):
+        """Calculate the expected predition in the untransfomred domain from ``pred``.
+        """
+        raise NotImplementedError()
+
+    def initialize_bias(self, bias: torch.Tensor):
```
Comment (breakds) [inline alf/utils/losses.py:126]: When reading `ScalarPredictionLoss` alone, I find it a bit clueless about why definition of the loss cares about initialization on the last FC layer of the prediction. Can we copy some of motivation from concrete class below here as example?

File: alf/utils/losses.py [hunk 2]
```diff
@@ -102,3 +104,371 @@ def multi_quantile_huber_loss(quantiles: torch.Tensor,
...
+
+        Args:
+            pred: raw model prediction
+        """
+        if self._transform is not None:
+            pred = self._transform.inverse_transform(pred)
+        return pred
+
+
+def _get_indexer(shape):
```
Comment (breakds) [inline alf/utils/losses.py]: It would be great to add documentation for this function.

Also I would have used "torch.meshgrid" without learning the broadcast trick from reading this implementation. It basically fulfills the same purpose as `torch.meshgrid` with linear memory consumption. Thanks!

File: alf/utils/losses.py [hunk 3]
```diff
@@ -102,3 +104,371 @@ def multi_quantile_huber_loss(quantiles: torch.Tensor,
...
+                 inverse_after_mean=False):
+        super().__init__()
+        self._transform = transform
+        if self._transform is not None:
+            self._inverse_after_mean = inverse_after_mean
+        else:
+            self._inverse_after_mean = True
+        self._support = None
+
+    def _calc_support(self, n):
```
Comment (breakds) [inline alf/utils/losses.py]: Nit: `n: int`.

**Review by le-horizon — DISMISSED** [review]
LG

File: alf/utils/losses.py [hunk 4]
```diff
@@ -102,3 +104,371 @@ def multi_quantile_huber_loss(quantiles: torch.Tensor,
...
+            return self._support
+        upper_bound = n // 2
+        lower_bound = -((n - 1) // 2)
+        x = torch.arange(lower_bound, upper_bound + 1, dtype=torch.float32)
+        if self._transform is not None and not self._inverse_after_mean:
+            x = self._transform.inverse_transform(x)
+        self._support = x
+        return x
+
+    def _calc_bin(self, logits, target):
```
Comment (breakds) [inline alf/utils/losses.py:229]: Suggesting adding documentation for this core function.

Comment (emailweixu) [reply alf/utils/losses.py:229]: added.

**Review by le-horizon — APPROVED** [review]
LG

**Review by breakds — APPROVED** [review]
Thanks!
